src/intro_movie_scenes_lib.c

```diff
@@ -0,0 +1,528 @@
+    }
+    ScheduleSetBgPosText(data->bgConfig, data->bgId, BG_POS_OP_SET_X, x);
+    ScheduleSetBgPosText(data->bgConfig, data->bgId, BG_POS_OP_SET_Y, y);
+}
+
+BOOL IntroMovie_WaitBgScrollAnim(IntroMovieBgScrollAnim *data, enum GFBgLayer bgId) {
+    BOOL ret = FALSE;
+    IntroMovieBgScrollAnim *obj = &data[IntroMovie_BgLayerToScrollEffectSlot(bgId)];
+    if (!obj->active) {
+        return TRUE;
+    }
+    if (obj->finished) {
+        ret = TRUE;
+    }
+    return ret;
+}
+
+void IntroMovie_CancelBgScrollAnim(IntroMovieBgScrollAnim *data, enum GFBgLayer bgId) {
+    IntroMovieBgScrollAnim *obj = &data[IntroMovie_BgLayerToScrollEffectSlot(bgId)];
+    if (obj->active) {
+        SysTask_Destroy(obj->task);
+        obj->task = NULL;
+        obj->finished = TRUE;
+        obj->active = FALSE;
+    }
+}
+
+int IntroMovie_BgLayerToScrollEffectSlot(enum GFBgLayer bgId) {
+    int ret = 0;
+    switch (bgId) {
+    case GF_BG_LYR_MAIN_0:
+        ret = 0;
+        break;
+    case GF_BG_LYR_MAIN_1:
+        ret = 1;
+        break;
+    case GF_BG_LYR_MAIN_2:
+        ret = 2;
+        break;
+    case GF_BG_LYR_MAIN_3:
+        ret = 3;
+        break;
+    case GF_BG_LYR_SUB_0:
+        ret = 4;
+        break;
+    case GF_BG_LYR_SUB_1:
+        ret = 5;
+        break;
+    case GF_BG_LYR_SUB_2:
+        ret = 6;
+        break;
+    case GF_BG_LYR_SUB_3:
+        ret = 7;
+        break;
+    default:
+        GF_ASSERT(FALSE);
+    }
+
+    return ret;
+}
+
+IntroMovieBgWindowAnim *IntroMovie_StartWindowPanEffect(IntroMovieBgWindowAnim *data, int duration, int whichScreen, const IntroMovieBgWindowAnimParam *param) {
+    IntroMovie_WindowsOn_SetInsideOutsidePlanes(param->winIn, param->winOut, param->topScreenEffect, param->bottomScreenEffect, whichScreen);
+    if (duration <= 0) {
+        IntroMovie_SetBgWindowsPosition(param->x1End, param->y1End, param->x2End, param->y2End, whichScreen);
+        return NULL;
+    }
+    IntroMovieBgWindowAnim *ret = whichScreen == 0 ? &data[1] : &data[0];
+    ret->param = *param;
+    ret->duration = duration;
+    ret->counter = 0;
+    ret->finished = TRUE;
+    ret->active = TRUE;
+    ret->whichScreen = whichScreen;
+    ret->x1 = param->x1Start;
+    ret->y1 = param->y1Start;
+    ret->x2 = param->x2Start;
+    ret->y2 = param->y2Start;
+    IntroMovie_SetBgWindowsPosition(param->x1Start, param->y1Start, param->x2Start, param->y2Start, whichScreen);
+    ret->task = SysTask_CreateOnVBlankQueue(Task_IntroMovie_WindowPanEffect, ret, 0);
+    return ret;
+}
+
+BOOL IntroMovie_WaitWindowPanEffect(IntroMovieBgWindowAnim *data, int a1) {
+    IntroMovieBgWindowAnim *which = a1 == 0 ? &data[1] : &data[0];
+    if (which->active == 0) {
+        return TRUE;
+    }
+    return which->finished != 0;
+}
+
+#define scalePos(start, end, pos, duration) ({ \
+    int diff = (end) - (start);            \
+    diff = diff * (pos) / (duration);          \
+    (start) + diff;                        \
```

Comment:
huh? how does this line work? is this a weird macro thing i dont understand?

Reply:
the result of the final expression is assigned to the lhs of a macro call

Reply:
gonna put it down to macro magic, and a C feature I don't fully understand